hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/utils/AzcopyToolHelper.java

```diff
@@ -0,0 +1,245 @@
+ * We will leverage this tool to create implicit paths in storage account for testing purposes.
+ */
+
+public class AzcopyToolHelper {
+
+  private File hadoopAzureDir;
+  private String azcopyDirPath;
+  private String accountName;
+  private String fileSystemName;
+  private String sasToken;
+
+  private static final String USER_DIR_SYSTEM_PROPERTY = "user.dir";
+  private static final String HADOOP_AZURE_DIR = "hadoop-azure";
+  private static final String AZCOPY_DIR_NAME = "azcopy";
+  private static final String AZCOPY_TOOL_FILE_NAME = "azcopy";
+  private static final String DIR_NOT_FOUND_ERROR = " directory not found";
+  private static final String AZCOPY_DOWNLOADED_DIR_NAME = "/azcopy_linux_amd64_*/* ";
+  private static final String AZCOPY_DOWNLOADED_TAR_NAME = "/azcopy_linux_amd64_* azcopy.tar.gz";
+
+  private static final String AZCOPY_CMD_SHELL = "bash";
+  private static final String AZCOPY_CMD_OPTION = "-c";
+  private static final String AZCOPY_DOWNLOAD_URL = "https://aka.ms/downloadazcopy-v10-linux";
+  private static final String AZCOPY_DOWNLOAD_CMD = "wget " + AZCOPY_DOWNLOAD_URL + " -O azcopy.tar.gz" + " --no-check-certificate";
+  private static final String EXTRACT_CMD = "tar -xf azcopy.tar.gz -C ";
+  private static final String MOVE_CMD = "mv ";
+  private static final String REMOVE_CMD = "rm -rf ";
+  private static final String CHMOD_CMD = "chmod +x ";
+
+  /**
+   * Azcopy tool work with SAS based authentication. SAS can be configured using
+   * test configuration "fs.azure.test.fixed.sas.token".
+   * @param accountName on which tool as has to run.
+   * @param fileSystemName container inside which implicit path has to be created.
+   * @param sasToken to be used by azcopy tool for authentication.
+   */
+  public AzcopyToolHelper(String accountName, String fileSystemName, String sasToken) {
+    this.accountName = accountName.replace(ABFS_DNS_PREFIX, WASB_DNS_PREFIX);
+    this.fileSystemName = fileSystemName;
+    this.sasToken = sasToken;
+  }
+
+  /**
+   * Create a file in the container using Azcopy tool.
+   * @param pathFromContainerRoot absolute path to be created.
+   * @throws Exception
+   */
+  public void createFileUsingAzcopy(String pathFromContainerRoot) throws Exception {
+    // Add the path you want to copy to as config.
+    if (pathFromContainerRoot != null) {
+      createFileOrFolder(pathFromContainerRoot, true);
+    }
+  }
+
+  /**
+   * Create a folder in the container using Azcopy tool.
+   * @param pathFromContainerRoot absolute path to be created.
+   * @throws Exception
+   */
+  public void createFolderUsingAzcopy(String pathFromContainerRoot) throws Exception {
+    // Add the path you want to copy to as config.
+    if (pathFromContainerRoot != null) {
+      createFileOrFolder(pathFromContainerRoot, false);
+    }
+  }
+
+  private void downloadAzcopyExecutableIfNotPresent() throws IOException, InterruptedException {
```

Comment:
lets have it synchronized; also, cache the binary path so all test in the jvm can use it.

Reply:
also, lets say second thread comes in this method, that should not execute anything and just return back.

Reply:
Changed the design. Please review the new code and let me know if this is still a problem.